### PR TITLE
add HACKING.md and move content from README.md over

### PR DESCRIPTION
**Pull request by mvo5**

### CODE REVIEW

**Comment by zyga** [comment]
+1

File: README.md [hunk 1]
```diff
@@ -1,162 +1,15 @@
...
-printed out something like `Listening on /run/snapd.socket as 3.` you
-should then
-
-    sudo chmod 0666 /run/snapd.socket
-
-so the socket has the right permissions (otherwise you need `sudo` to
-connect).
+See
+[HACKING.md](https://github.com/snapcore/snapd/blob/master/HACKING.md)
+for in-depth details how to hack on snapd.
```
Comment (chipaca) [inline README.md]: > in-depth details *on* how to hack on snapd.

But if the two *on*s bother you maybe rephrase it entirely.

File: HACKING.md [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+integration test framework for the integration/system level tests.
+
+## Development
+
+### Setting up a GOPATH
+
+When working with the source of Go programs, you should define a path within
+your home directory (or other workspace) which will be your `GOPATH`. `GOPATH`
+is similar to Java's `CLASSPATH` or Python's `~/.local`. `GOPATH` is documented
+[http://golang.org/pkg/go/build/](online) and inside the go tool itself
```
Comment (chipaca) [inline HACKING.md]: I know you're just moving stuff around, but while you're here: that link isn't. I believe the intent was to have it the other way around, ie a link that said "online" and pointed you to that URL. It needs reversing: `[online](http://golang.org/pkg/go/build/)`.

File: HACKING.md [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+Setting `GOPATH` correctly is critical when developing Go programs. Set and
+export it as part of your login script.
+
+Add `$GOPATH/bin` to your `PATH`, so you can run the go programs you install:
+
+    PATH="$PATH:$GOPATH/bin"
+
+### Getting the snappy sources
+
+The easiest way to get the source for `snappy` is to use the `go get` command.
```
Comment (chipaca) [inline HACKING.md]: we call it `snapd` these days :-)

File: HACKING.md [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+
+    go get -d -v github.com/snapcore/snapd/...
+
+This command will checkout the source of `snappy` and inspect it for any unmet
```
Comment (chipaca) [inline HACKING.md]: snapd

File: HACKING.md [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+Go package dependencies, downloading those as well. `go get` will also build
+and install `snappy` and its dependencies. To checkout without installing, use
```
Comment (chipaca) [inline HACKING.md]: snapd.

This paragraph seems to assume the command we just told the user to do was without `-d`; I think it needs reversing (e.g “to also build and install `snapd` itself into `${GOPATH}/bin`, omit the `-d` flag”)

File: HACKING.md [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+the `-d` flag. More details on the `go get` flags are available using
+
+    go help get
+
+At this point you will have the git local repository of the `snappy` source at
```
Comment (chipaca) [inline HACKING.md]: `snapd`

File: HACKING.md [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+
+### Dependencies handling
+
+To generate dependencies.tsv you need `godeps`, so
+
+    go get launchpad.net/godeps
+
+To obtain the correct dependencies for the project, run:
+
+    godeps -t -u dependencies.tsv
```
Comment (chipaca) [inline HACKING.md]: this needs to be run from the project root i believe

File: HACKING.md [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+
+You can use the script `get-deps.sh` to run the two previous steps.
+
+If the dependencies need updating
+
+    godeps -t ./... > dependencies.tsv
```
Comment (chipaca) [inline HACKING.md]: as does this

File: HACKING.md [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+Canonical contributor licence agreement at
+http://www.ubuntu.com/legal/contributors
+
+Snappy can be found on Github, so in order to fork the source and contribute,
+go to https://github.com/snapcore/snapd. Check out [Github's help
+pages](https://help.github.com/) to find out how to set up your local branch,
+commit changes and create pull requests.
+
+We value good tests, so when you fix a bug or add a new feature we highly
+encourage you to create a test in $source_testing.go. See also the section
```
Comment (chipaca) [inline HACKING.md]: backticks around `$souorce_testing.go`

File: HACKING.md [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+about Testing.
+
+### Testing
+
+To run the various tests that we have to ensure a high quality source just run:
+
+    ./run-checks
+
+This will check if the source format is consistent, that it build, all tests
```
Comment (chipaca) [inline HACKING.md]: that it build*s*

File: HACKING.md [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+work as expected and that "go vet" and "golint" have nothing to complain.
```
Comment (chipaca) [inline HACKING.md]: we don't really block on golint (and it complains a lot these days)

File: HACKING.md [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+
+There is more to read about the testing framework on the [website](https://labix.org/gocheck)
+
+### Running the spread tests
+
+To run the spread tests locally you need the latest version of spread
+from https://github.com/snapcore/spread. It can be installed via:
+
+    $ sudo apt install qemu-kvm autopkgtest
+    $ sudo snap install spread
```
Comment (chipaca) [inline HACKING.md]: not sure what to do about this aprt, because the spread snap doesn't work

Comment (chipaca) [inline HACKING.md]: (and the instructions as given are wrong even if it were to work)

File: HACKING.md [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+    $ spread -keep qemu:
+
+It will print how to reuse the systems. Make sure to use
+`export REUSE_PROJECT=1` in your environment too.
+
+
+### Testing snapd on a snappy system
+
+To test the `snapd` REST API daemon on a snappy system you need to
+transfer it to the snappy system and then run:
```
Comment (chipaca) [inline HACKING.md]: you can also test it this way on classic, fwiw

File: HACKING.md [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+
+    sudo systemctl stop snapd.service snapd.socket
+    sudo /lib/systemd/systemd-activate -E SNAPD_DEBUG=1 -l /run/snapd.socket -l /run/snapd-snap.socket ./snapd
```
Comment (chipaca) [inline HACKING.md]: does this need `SNAP_REEXEC=0`?

Comment (chipaca) [inline HACKING.md]: also probably good to have `SNAPD_DEBUG_HTTP=3`

File: HACKING.md [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+
+This will stop the installed snapd and activate the new one. Once it's
+printed out something like `Listening on /run/snapd.socket as 3.` you
+should then
+
+    sudo chmod 0666 /run/snapd.socket
```
Comment (chipaca) [inline HACKING.md]: make that `snapd*.socket` so it also chmods `snapd-snap.socket`

**Comment by mvo5** [comment]
Thanks for the detailed review @chipaca! I addressed the feedback now.
